overlord/snapshotstate/backend/backend_test.go

```diff
@@ -324,26 +347,120 @@ func (s *snapshotSuite) TestIterReportsCloseError(c *check.C) {
 	c.Check(calledF, check.Equals, true)
 }
 
+func readerForFilename(fname string, c *check.C) *backend.Reader {
+	var snapname string
+	var id uint64
+	fn := strings.TrimSuffix(filepath.Base(fname), ".zip")
+	_, err := fmt.Sscanf(fn, "%d_%s", &id, &snapname)
+	c.Assert(err, check.IsNil, check.Commentf(fn))
+	f, err := os.Open(os.DevNull)
+	c.Assert(err, check.IsNil, check.Commentf(fn))
+	return &backend.Reader{
+		File: f,
+		Snapshot: client.Snapshot{
+			SetID: id,
+			Snap:  snapname,
+		},
+	}
+}
+
+func (s *snapshotSuite) TestIterIgnoresSnapshotsWithSetIdMismatches(c *check.C) {
+	logbuf, restore := logger.MockLogger()
+	defer restore()
+
+	defer backend.MockOsOpen(func(string) (*os.File, error) {
+		return new(os.File), nil
+	})()
+	readNames := 0
+	defer backend.MockDirNames(func(*os.File, int) ([]string, error) {
+		readNames++
+		if readNames > 2 {
+			return nil, io.EOF
+		}
+		return []string{
+			"1_foo.zip",
+			"2_bar.zip"}, nil
+	})()
+	defer backend.MockOpen(func(fname string) (*backend.Reader, error) {
+		r := readerForFilename(fname, c)
+		if r.SetID == 1 {
+			r.SetID = 99
+		}
+		return r, nil
+	})()
+
+	calledF := false
+	f := func(snapshot *backend.Reader) error {
+		calledF = true
+		c.Check(snapshot.SetID, check.Equals, uint64(2))
+		return nil
+	}
+
+	err := backend.Iter(context.Background(), f)
+	c.Check(err, check.IsNil)
+	c.Check(logbuf.String(), check.Matches, `(?m).* Snapshot \"1_foo.zip\" ignored, internal set-id 99 disagrees with filename`)
+	c.Check(calledF, check.Equals, true)
+}
+
+func (s *snapshotSuite) TestIterIgnoresSnapshotsWithInvalidNames(c *check.C) {
+	logbuf, restore := logger.MockLogger()
+	defer restore()
+
+	defer backend.MockOsOpen(func(string) (*os.File, error) {
+		return new(os.File), nil
+	})()
+	readNames := 0
+	defer backend.MockDirNames(func(*os.File, int) ([]string, error) {
+		readNames++
+		if readNames > 4 {
```

Comment:
This should be "readNames > 1", the code inbackend.go is a bit confusing, see my comment above/below about using a counter for "calledF". dirNames() will be called in a loop but all the relevant names are already returned in the first loop so no need to run it twice.